Modules/_threadmodule.c

```diff
@@ -22,7 +22,243 @@
 static struct PyModuleDef thread_module;
 
 
+/* threads owned by the modulo */
+
+struct module_thread {
+    PyThreadState *tstate;
+    int daemonic;
+    PyThread_type_lock lifetime_mutex;
+    int lifetime_mutex_held;
+    struct module_thread *prev;
+    struct module_thread *next;
+};
+
+struct module_threads {
+    PyThread_type_lock mutex;
+    struct module_thread *head;
+    struct module_thread *tail;
+};
+
+static int
+module_threads_init(struct module_threads *threads)
+{
+    threads->head = NULL;
+    threads->tail = NULL;
+    threads->mutex = PyThread_allocate_lock();
+    if (threads->mutex == NULL) {
+        PyErr_NoMemory();
+        return -1;
+    }
+    return 0;
+}
+
+#ifdef HAVE_FORK
+static int module_thread_reinit(struct module_thread *);
+
+static int
+module_threads_reinit(struct module_threads *threads)
+{
+    if (_PyThread_at_fork_reinit(threads->mutex) < 0) {
+        PyErr_SetString(ThreadError, "failed to reinitialize lock at fork");
+        return -1;
+    }
+
+    PyThread_acquire_lock(threads->mutex, WAIT_LOCK);
+
+    struct module_thread *mt = threads->head;
+    while (mt != NULL) {
+        if (module_thread_reinit(mt) < 0) {
+            return -1;
+        }
+        mt = mt->next;
+    }
+
+    PyThread_release_lock(threads->mutex);
+
+    return 0;
+}
+#endif
+
+static void
+module_threads_fini(struct module_threads *threads)
+{
+    // Wait for all the threads to finalize.
+    PyThread_acquire_lock(threads->mutex, WAIT_LOCK);
+    while (threads->head != NULL) {
+        PyThread_release_lock(threads->mutex);
+        // XXX Sleep?
+        PyThread_acquire_lock(threads->mutex, WAIT_LOCK);
+    }
+    PyThread_release_lock(threads->mutex);
```

Comment:
FYI, this doesn't feel right.  I'm going to look at what my options are.  A condvar would work, but that doesn't seem to be something we do in CPython outside the GIL.

(I'm also open to ideas.)

Reply:
I've come up with a slightly better approach.